### ISSUE TITLE
should have comment or be unexported

### ISSUE THREAD

**Issue by agjs** [golang/lint#191]
<img width="798" alt="screen shot 2016-02-03 at 18 20 59" src="https://cloud.githubusercontent.com/assets/7226555/12790524/ee0ce714-caa2-11e5-862c-afffdb4feec5.png">

I've just started learning Golang and while I do want to follow best practises and follow this linter, I'm not sure how to fix these warnings. When I put some random comment above some function or a type nothing really happens.

For instance, if I put a comment above "type Saturation struct" I get another warning message :

> comment on exported type Salutation should be of the form "Salutation ..." (with optional leading article)

Please if you can, guide me on how to comment my code properly so I can avoid these warnings, they are freaking me out like hell.

**Comment by dsymonds** [comment]
https://golang.org/s/style is probably a good place to start reading, including the places it links to.

**Comment by agjs** [comment]
Can you just answer the question instead of sending a link to entire documentation like a generic robot ?

**Comment by dsymonds** [comment]
For that specific single lint message, the bit under "Doc Comments" is what you want. But you'll want the whole doc to understand other messages that lint produces too.

**Comment by alvarolm** [comment]
While I to belive that following the guidelines is the best practice, its usefull to remeber that this kind of features could be implemented on top of the linter as this commit states: https://github.com/golang/lint/pull/192/commits/37c4c6fcefd668054d6430b741ee64f1e826cda0

**Comment by mairanteodoro** [comment]
Hi, @agjs !

I don't know if you're still having this problems with this issue, but the solution to it is to add a comment right before anything that you export. For example, in your case, you could simply do this:

`// Salutation : here you tell us what Salutation is`
`// Printer : what is this?`
`// Greet : describe what this function does`
`// CreateMessage : describe what this function does`

And so on. Note that the colon is optional but there _must_ be a blank space between the first word (the name of the thing that's being exported) and the next one (the description of what that thing is).

The point is that _**the comment has to start with the name of the thing that's being exported**_.

**Comment by lsianturi** [comment]
Here is an example, for whoever who might start learning Go

// Vertex blablabla
type Vertex struct {
	X, Y float64
}

// Abs blablaba
func (v Vertex) Abs() float64 {
	return math.Sqrt(v.X*v.X + v.Y*v.Y)
}

**Comment by Tower450** [comment]
thanks

**Comment by linzzz0** [comment]
Got it!

**Comment by Godlovesmile** [comment]
Just add a meaningful comment to the struct

// Person is a representation of a person
type Person struct {
    Name string
    Phone string
}

**Comment by Godlovesmile** [comment]
Just add a meaningful comment to the struct

// Person is a representation of a person
type Person struct {
        Name string
        Phone string
}

**Comment by Godlovesmile** [comment]
Just add a meaningful comment to the struct (or the func)

// Person is a representation of a person
type Person struct {
   Name string
   Phone string
}

**Comment by gocs** [comment]
Why does it warn on a multiline comment?

**Comment by powdevel** [comment]
Answering every question that is already documented may seem redundant. However, when a person is stuck on something a simple answer can help a lot. This seems a use case for a chat bot that would provide quick help for most simple and direct questions. I usually read docs when I feel comfortable on a topic, not at a beginner level.
Thanks @mairanteodoro @lsianturi now I can go back learning _Go_.

**Comment by liuchong** [comment]
> Can you just answer the question instead of sending a link to entire documentation like a generic robot ?

666

**Comment by ulidtko** [comment]
I'm just writing a first program and trying the tools, why do I have to document anything at all?..

Please explain the **"be unexported"** part, too. I don't intend to export anything from `package main`.

**Comment by mrzhouxiaofei** [comment]
> Hi, @agjs !
> 
> I don't know if you're still having problems with this issue, but the solution to it is to add a comment right before anything that you export. For example, in your case, you could simply do this:
> 
> `// Salutation : here you tell us what Salutation is`
> `// Printer : what is this?`
> `// Greet : describe what this function does`
> `// CreateMessage : describe what this function does`
> 
> And so on. Note that the colon is optional but there _must_ be a blank space between the first word (the name of the thing that's being exported) and the next one (the description of what that thing is).
> 
> The point is that _**the comment has to start with the name of the thing that's being exported**_.

Thanks

**Comment by mrzhouxiaofei** [comment]
Thanks，I get it.

**Comment by mouryavenkat** [comment]
// YourHandler ... This function is just to handle the incoming request
func YourHandler(w http.ResponseWriter, r *http.Request) {
	w.Write([]byte("Gorilla!\n"))
}

Apparently GO expects to have a comment line on top of every function in a specific format. 
The format is 

// MethodName ... < Some lines that describe your function>

Hope this helps.

**Comment by goldyfruit** [comment]
> it warns in like this
> 
> ```
> /** Public is a public function.
>  *  usage: ...
>  */
> ```
> 
> among other cases that uses `/**/`
> 
> same question I had

Got the same behavior, with `/** */` I got complains but if I switch to `//` then it's fine.

**Comment by viniciusbds** [comment]
thanks @mouryavenkat .

**Comment by layzbot** [comment]
Although this is closed, I'd just like to ask one thing. This suggestion pops up when I'm starting struct with uppercase like 'type Person struct....' but is magically removed when creating it with small case first character like 'type person struct...'
Why so?

**Comment by mouryavenkat** [comment]
@layzbot  The reason is when you start a function name with capital letter, it means the function is to be exported. For any exported function there should be a comment line.
But thats not the case with functions starting with small letters.

**Comment by layzbot** [comment]
Oh! Thanks @mouryavenkat!

**Comment by gocpplua** [comment]
@mouryavenkat thanks

**Comment by rzr** [comment]
May I suggest a feature that would fix code and eventually insert TODO hints inline

**Comment by SLCH** [comment]
Multiline comments should be valid.

**Comment by King-Success** [comment]
Great, saves me a lot of time

**Comment by nishant8BITS** [comment]
> Can you just answer the question instead of sending a link to entire documentation like a generic robot ?

Honest reply!

**Comment by shayan09** [comment]
Add a comment before the start of the function. The comment should start with the name of the function.
Something like,
```
// foo is an example function
func foo() {
---- CODE----
}

**Comment by joooni1998** [comment]
So... the purpose of the error is to make you explain what are you gonna do with the exported function?
GOT IT

**Comment by codemobiles** [comment]
Here is an example to fix the warning.  By commenting above the function with this format
// MethodName : <your comment>

// GetGoldData : Handle goldata path
func GetGoldData(c *gin.Context){	
	c.Data(200, "application/json; charset=utf-8", []byte(GoldData))	
}


// GetExchangeRate : Handle exchangerate path
func GetExchangeRate(c *gin.Context){	
	c.Data(200, "application/json; charset=utf-8", []byte(ExchangeRate))	
}

**Comment by victorbiga** [comment]
> Add a comment before the start of the function. The comment should start with the name of the function.
> Something like,
> 
> ```
> // foo is an example function
> func foo() {
> ---- CODE----
> }
> ```

this is the exact answer to this issue

**Comment by aadamowski** [comment]
How about functions that are exported to C ?

Per https://golang.org/cmd/cgo/#hdr-C_references_to_Go , you do it by labeling the functions with comments of the following form:

```
//export MyFunction
func MyFunction(arg1, arg2 int, arg3 string) int64 {...}
```
This, however, conflicts with the discussed lint rule. For example, this Go implementation of a PKCS#11 API function exported to C:

```
//export CGOGetTokenInfo
func CGOGetTokenInfo(slotID C.CK_SLOT_ID, pInfo C.CK_TOKEN_INFO_PTR) C.CK_RV {
```
It provokes this complaint from lint:

```
comment on exported function CGOGetTokenInfo should be of the form "CGOGetTokenInfo ..."
```

To the best of my knowledge, there is no documented right way to satisfy this lint rule in such a case - https://github.com/golang/go/wiki/CodeReviewComments#doc-comments and https://golang.org/doc/effective_go.html#commentary make no acknowledgement of the special C export comment syntax.

**Comment by ulidtko** [comment]
@aadamowski that looks like a genuine bug. You may have better luck with opening a separate issue for it (because this one is so crowded, and closed since 4 years ago).

**Comment by ianlancetaylor** [comment]
@aadamowski This issue is closed.  Commenting on closed issues won't get you far.

In any case we recommend writing comments like

```Go
// CGOGetTokenInfo is called from C code to get token info.
//export CGOGetTokenInfo
func CGOGetTokenInfo(...)
```
